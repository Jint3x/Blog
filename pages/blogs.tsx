import React, { useState } from "react"
import useSWR from 'swr'

import Header from "../components/client-side/Header"
import Footer from "../components/client-side/Footer"
import BlogPostPreview from "../components/client-side/PostPreview"

const fetcher = async function<JSON = any>(
    input: RequestInfo,
    init?: RequestInit
  ): Promise<JSON> {
    const res = await fetch(input, init)
    return res.json()
  }

interface PostPreview {
    thumbnail: string,
    title: string,
    description: string,
    date: string,
    tags: string[],
    url: string,
}


interface ListOfBlogs_I {
    blogPosts: PostPreview[]
}
 
function ListOfBlogs(props: ListOfBlogs_I) {
    return (
        <>
        <Header />
        <BlogPage blogs={props.blogPosts} />
        <Footer />
        </>
    )
}


interface BlogPage_I {
    blogs: PostPreview[]
}

function BlogPage(props: BlogPage_I) {
    const [posts, setPosts] = useState(props.blogs);
    const [page, setPage] = useState(1);

    const {data, isValidating, error} = useSWR(`/api/blog?page=${page}&amount=${6}`, fetcher, {initialData: props.blogs, revalidateOnFocus: false, revalidateOnMount: false})

    return (
        <section>
            <div className={"pt-24 pb-12 sm:pt-16"}>
                <FilterBlogs />

                <DisplayBlogs 
                 isLoading={false}
                 blogs={page === 1 ? data as PostPreview[] : []} 
                />

                <BlogPagination 
                 page={page} 
                 setPage={setPage}
                 amountOfPosts={posts.length}
                />
            </div>
        </section>
    )
}


function FilterBlogs() {
    return (
        <div>
            <h1 className={"text-center text-2xl"}>Search Blogs</h1>

            <input 
             type={"text"}
             className={"w-64 sm:w-80 md:w-96 lg:w-128 h-6 sm:h-8 border pl-2 pr-2 bg-main-green-trans border-2 border-solid border-main-green rounded ml-auto mr-auto block mt-4 outline-none"} 
            />
            <p className={"text-red-500 w-64 md:w-96 lg:w-128 ml-auto mr-auto text-center"}>Search feature not available yet</p>
        </div>
    )
}


interface DisplayBlogs_I {
    blogs: PostPreview[],
    isLoading: boolean,
}
// Take url props to fetch
function DisplayBlogs(props: DisplayBlogs_I) {

    return (
        <section>
            <div className={`flex sm:max-w-3xl flex-col flex-wrap justify-space-around ml-auto mr-auto xl:max-w-5xl sm:min-w-sm ${props.isLoading || props.blogs.length < 1 ? "sm:flex" : "sm:grid"} sm:grid-cols-2 sm:grid-rows-${Math.ceil(props.blogs.length / 2)} xl:grid-cols-3 xl:grid-rows-2`}> 
                {
                    props.isLoading ? 
                    <div className={"mt-32 mb-32 sm:mt-48 sm:mb-48 md:mb-48 lg:mb-64 lg:mt-64 animate-spinner ml-auto mr-auto ease-linear rounded-full border-8 border-gray-200 h-24 w-24"} />
                    :
                    props.blogs.length < 1 ?
                    <div>
                        <h2 className={"text-center mt-32 mb-32 text-xl sm:mt-64 sm:mb-64 xl:mt-80 xl:mb-80"}>No blogs available</h2>
                    </div>
                    :
                    (props.blogs.map(blog =>
                        <BlogPostPreview 
                         thumbnail={blog.thumbnail}
                         title={blog.title}
                         description={blog.description}
                         date={blog.date}
                         tags={blog.tags}
                         url={blog.url}
                         key={blog.title}
                        />
                    ))
                }
                
            </div>
        </section>
    )
}

interface BlogPagination_I {
    page: number,
    setPage: React.Dispatch<React.SetStateAction<number>>,
    amountOfPosts: number
}

function BlogPagination(props: BlogPagination_I) {
    return (
        <div>
            <p className={"w-12 flex justify-around ml-auto mr-auto"}>
                <span 
                 className={"cursor-pointer"}
                 onClick={() => props.setPage(oldpage => oldpage - 1)}
                >
                    {props.page === 1 ? "" : props.page - 1}
                </span>

                <span className={"text-main-green"}>
                    {props.page}
                </span>

                <span 
                 className={"cursor-pointer"}
                 onClick={() => props.setPage(oldpage => oldpage + 1)}
                >
                    {props.amountOfPosts === 6 ? props.page + 1 : ""}
                </span>
            </p>
        </div>
    )
}



export async function getStaticProps() {
    // Fetch the latest 6 posts from here with a fetch function

    let latestPost = {
        thumbnail: "/images/homepage/blogs-introduction.jpg",
        title: "A title",
        description: 
        `As a developer, I’ve always asked myself ‘how?’. In a world, where things just keep getting more complex and difficult to learn, I want to create a place,
        a community where even the most complex things about technology and programming are explored in depth and fundametally understood. The first step towards
        achieving this goal, is this website. Here, I will try to go very in depth and explain many and different subjects about computers and programming.`, 
        date: "2nd of Feb, 2021",
        tags: ["Deep Dive", "SDF", "Tdsogy", "Technology", "Technology"],
        url: "/blog/how-to-test"
    }

    return {
        revalidate: 1800,
        props: { blogPosts: [latestPost, latestPost, latestPost, latestPost, latestPost, latestPost] }
    }
}

export {FilterBlogs, DisplayBlogs, BlogPagination}
export default ListOfBlogs