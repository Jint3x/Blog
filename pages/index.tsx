import Header from "../components/client-side/Header"
import Footer from "../components/client-side/Footer"
import PostPreview from "../components/client-side/PostPreview"

import React, { Dispatch, SetStateAction, useState } from "react"
import { GetStaticProps, GetServerSidePropsContext } from 'next'



interface Homepage_I {
    latestPost: PostPreviewData
}

function Homepage(props: Homepage_I) {
    return (
        <div>
            <Header />
            <BlogIntroduction />
            <LatestPost recentPost={props.latestPost} />
            <Footer />
        </div>
    )
}


function BlogIntroduction() { // Commit later
    return (
        <section>
            <div className={"w-full bg-homepage-introduction pt-24 text-default-white bg-center"}>
                <h1 className={"text-center text-3xl"}>
                    <span className={"text-main-blue relative bottom-2 right-1"}>Jint3x's</span>
                    <span className={"text-main-green"}>Blogs</span>
                </h1>

                <h2 className={"text-center text-xl"}>
                    Satisfy your curiosity
                </h2>

                <p className={"w-4/5 text-center pt-4 pb-24 ml-auto mr-auto sm:w-128 sm:text-lg lg:w-192"}>
                    As a developer, I’ve always asked myself ‘how?’. In a world, where things just keep getting more complex and difficult to learn, I want to create a place,
                    a community where even the most complex things about technology and programming are explored in depth and fundametally understood. The first step towards
                    achieving this goal, is this website. Here, I will try to go very in depth and explain many and different subjects about computers and programming.
                </p>
            </div>
        </section>
    )
}

interface PostPreviewData {
    thumbnail: string,
    title: string,
    description: string,
    date: string,
    tags: string[],
    url: string,
}

interface LatestPost_I {
    recentPost: PostPreviewData
}

function LatestPost(props: LatestPost_I) {
    const [postData, setPostData]: [PostPreviewData, Dispatch<SetStateAction<PostPreviewData>>]  = useState(props.recentPost)

    return (
        <div className={"max-w-lg ml-auto mr-auto"}>
            <h2 className={"text-center mt-12 text-2xl font-normal"}>Recently Posted Blogs</h2>

            <div className={"flex align-center justify-center flex-col sm:flex-row sm:items-center"}>
                <ClickButton extraClasses={"relative top-8 sm:static"} direction={"up"} />

                <div className={"relative"}>
                    <PostPreview 
                    thumbnail={postData.thumbnail}
                    title={postData.title}
                    description={postData.description}
                    date={postData.date}
                    tags={postData.tags}
                    url={postData.url}
                    />

                    <div className={"absolute w-72 h-80 bg-main-green transform translate-x-4 -rotate-6 top-0 -z-10 mt-12 mb-12 inset-x-1/2 -translate-x-1/2"} />
                 </div>

                <ClickButton extraClasses={"relative bottom-10 sm:static"} direction={"down"} />
            </div>
        </div>
    )
}

type ClickButtonDir = "up" | "down";

interface ClickButton_I {
    extraClasses?: string,
    direction: ClickButtonDir
}


// Need to realign the arrows when the screen becomes big enough
function ClickButton(props: ClickButton_I) {
    return (
        <div className={`w-12 h-12 rounded-full bg-black ml-auto mr-auto mt-2 cursor-pointer flex items-center ${props.extraClasses} transform duration-300 ${props.direction === "up" ? "sm:hover:-translate-x-2" : "sm:hover:translate-x-2"}`}>
            <img
             src={"/images/homepage/arrow.png"}
             alt={"Switch Blog Icon"}
             className={`w-8 block ml-auto mr-auto transform ${props.direction === "up" ? "-rotate-90 sm:rotate-180" : "rotate-90 sm:rotate-0"}`} 
            />
        </div>
    )
}


/**
 * TODO: 
 * Implement a blog fetch function which should be used when one of the arrows from above is
 * clicked (which will cause a new blog to be shown).
*/


export { BlogIntroduction, LatestPost }
export default Homepage


export async function getStaticProps(context: GetServerSidePropsContext) {
    // Fetch the latest post from here 

    let latestPost = {
        thumbnail: "/images/homepage/blogs-introduction.jpg",
        title: "A title",
        description: 
        `As a developer, I’ve always asked myself ‘how?’. In a world, where things just keep getting more complex and difficult to learn, I want to create a place,
        a community where even the most complex things about technology and programming are explored in depth and fundametally understood. The first step towards
        achieving this goal, is this website. Here, I will try to go very in depth and explain many and different subjects about computers and programming.`, 
        date: "2nd of Feb, 2021",
        tags: ["Deep Dive", "SDF", "Tdsogy", "Technology", "Technology"],
        url: "/blog/how-to-test"
    }

    return {
        revalidate: 900,
        props: { latestPost }
    }
}