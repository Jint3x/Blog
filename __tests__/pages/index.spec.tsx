/**
 * @jest-environment jsdom
 */

import ReactRenderer from "react-test-renderer"
import { screen, render } from "@testing-library/react"
import { BlogIntroduction, LatestPost } from "../../pages/index"

test("Take a snapshot of the BlogIntroduction component", () => {
    let tree = ReactRenderer.create(<BlogIntroduction />).toJSON();
    expect(tree).toMatchSnapshot();
});


test("Render a Blog element and verify its items", () => {
    render(
        <LatestPost 
         recentPost={{
            thumbnail: "/images/homepage/blogs-introduction.jpg",
            title: "A title",
            description:  "As a developer, I’ve always asked myself ‘how?’. In a world....", 
            date: "2nd of Feb, 2021",
            tags: ["Deep Dive", "SDF", "Tdsogy", "Technology"],
            url: "/blog/how-to-test"
         }}
        />
    )

    let recent = screen.getByText("Recently Posted Blogs");
    let postContainer = recent.nextSibling as HTMLElement;


    expect(recent.nodeName).toBe("H2");
    expect(postContainer.childElementCount).toBe(3);
})

// Need to implement the fetching function first.
test.todo("Test the arrow functions")


test("Take a snapshot of the recently posted blog component", () => {
    let tree = ReactRenderer.create(
        <LatestPost 
         recentPost={{
            thumbnail: "/images/homepage/blogs-introduction.jpg",
            title: "A title",
            description:  "As a developer, I’ve always asked myself ‘how?’. In a world....", 
            date: "2nd of Feb, 2021",
            tags: ["Deep Dive", "SDF", "Tdsogy", "Technology"],
            url: "/blog/how-to-test"
         }}
        />
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
})