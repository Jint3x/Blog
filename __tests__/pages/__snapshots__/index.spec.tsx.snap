// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Take a snapshot of the BlogIntroduction component 1`] = `
<section>
  <div
    className="w-full bg-homepage-introduction pt-24 text-default-white bg-center"
  >
    <h1
      className="text-center text-3xl"
    >
      <span
        className="text-main-blue relative bottom-2 right-1"
      >
        Jint3x's
      </span>
      <span
        className="text-main-green"
      >
        Blogs
      </span>
    </h1>
    <h2
      className="text-center text-xl"
    >
      Satisfy your curiosity
    </h2>
    <p
      className="w-4/5 text-center pt-4 pb-24 ml-auto mr-auto sm:w-128 sm:text-lg lg:w-192"
    >
      As a developer, I’ve always asked myself ‘how?’. In a world, where things just keep getting more complex and difficult to learn, I want to create a place, a community where even the most complex things about technology and programming are explored in depth and fundametally understood. The first step towards achieving this goal, is this website. Here, I will try to go very in depth and explain many and different subjects about computers and programming.
    </p>
  </div>
</section>
`;

exports[`Take a snapshot of the recently posted blog component 1`] = `
<div
  className="max-w-lg ml-auto mr-auto"
>
  <h2
    className="text-center mt-12 text-2xl font-normal"
  >
    Recently Posted Blogs
  </h2>
  <div
    className="flex align-center justify-center flex-col sm:flex-row sm:items-center"
  >
    <div
      className="w-12 h-12 rounded-full bg-black ml-auto mr-auto mt-2 cursor-pointer flex items-center relative top-8 sm:static transform duration-300 sm:hover:-translate-x-2"
    >
      <img
        alt="Switch Blog Icon"
        className="w-8 block ml-auto mr-auto transform -rotate-90 sm:rotate-180"
        src="/images/homepage/arrow.png"
      />
    </div>
    <div
      className="relative"
    >
      <div
        className="w-72 bg-white ml-auto mr-auto h-80 rounded shadow-lg shadow shadow-md shadow-personal mt-12 mb-12 overflow-hidden"
      >
        <div
          className="h-20per bg-center overflow-hidden cursor-pointer"
          onClick={[Function]}
          style={
            Object {
              "backgroundImage": "url(/images/homepage/blogs-introduction.jpg)",
            }
          }
        />
        <div
          className="h-65per color-default-dark relative"
        >
          <h3
            className="text-center pt-1 pb-2 text-2xl overflow-hidden"
          >
            A title
          </h3>
          <div
            className="relative h-3/6 pl-2 pr-2 mt-2"
          >
            <p
              className="overflow-hidden h-full"
            >
              As a developer, I’ve always asked myself ‘how?’. In a world....
            </p>
            <div
              className="absolute bottom-0 h-12 w-full bg-gradient-to-t from-white to-transparent"
            />
          </div>
          <p
            className="absolute bottom-1 text-sm left-1 text-post-date"
          >
            2nd of Feb, 2021
          </p>
        </div>
        <div
          className="h-15per bg-main-blue-lighter flex-nowrap flex items-center overflow-hidden relative cursor-pointer"
          onClick={[Function]}
        >
          <div
            className="absolute h-full w-1/12 right-0 bg-gradient-to-l from-main-blue-lighter to-transparent"
          />
          <p
            className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
          >
            Deep Dive
          </p>
          <p
            className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
          >
            SDF
          </p>
          <p
            className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
          >
            Tdsogy
          </p>
          <p
            className="bg-main-green-trans ml-2 mr-2 text-sm rounded-md p-1 no-break whitespace-nowrap"
          >
            Technology
          </p>
        </div>
      </div>
      <div
        className="absolute w-72 h-80 bg-main-green transform translate-x-4 -rotate-6 top-0 -z-10 mt-12 mb-12 inset-x-1/2 -translate-x-1/2"
      />
    </div>
    <div
      className="w-12 h-12 rounded-full bg-black ml-auto mr-auto mt-2 cursor-pointer flex items-center relative bottom-10 sm:static transform duration-300 sm:hover:translate-x-2"
    >
      <img
        alt="Switch Blog Icon"
        className="w-8 block ml-auto mr-auto transform rotate-90 sm:rotate-0"
        src="/images/homepage/arrow.png"
      />
    </div>
  </div>
</div>
`;
